crates/milli/src/update/facet/new_incremental.rs

```diff
@@ -0,0 +1,477 @@
+    delta_data: Vec<FacetFieldIdChange>,
+}
+
+struct FacetsUpdateIncrementalInner {
+    db: Database<FacetGroupKeyCodec<BytesRefCodec>, FacetGroupValueCodec>,
+    field_id: FieldId,
+    group_size: u8,
+    min_level_size: u8,
+    max_group_size: u8,
+}
+
+impl FacetsUpdateIncremental {
+    pub fn new(
+        index: &Index,
+        facet_type: FacetType,
+        field_id: FieldId,
+        delta_data: Vec<FacetFieldIdChange>,
+        group_size: u8,
+        min_level_size: u8,
+        max_group_size: u8,
+    ) -> Self {
+        FacetsUpdateIncremental {
+            inner: FacetsUpdateIncrementalInner {
+                db: match facet_type {
+                    FacetType::String => index
+                        .facet_id_string_docids
+                        .remap_key_type::<FacetGroupKeyCodec<BytesRefCodec>>(),
+                    FacetType::Number => index
+                        .facet_id_f64_docids
+                        .remap_key_type::<FacetGroupKeyCodec<BytesRefCodec>>(),
+                },
+                field_id,
+                group_size,
+                min_level_size,
+                max_group_size,
+            },
+
+            delta_data,
+        }
+    }
+
+    #[tracing::instrument(level = "trace", skip_all, target = "indexing::facets::incremental")]
+    pub fn execute(mut self, wtxn: &mut RwTxn) -> Result<()> {
+        if self.delta_data.is_empty() {
+            return Ok(());
+        }
+        self.delta_data.sort_unstable_by(
+            |FacetFieldIdChange { facet_value: left, .. },
+             FacetFieldIdChange { facet_value: right, .. }| left.cmp(right),
+        );
+
+        self.inner.find_touched_parents(
+            wtxn,
+            0,
+            self.delta_data
+                .into_iter()
+                // reverse lexicographic order
+                .rev()
+                .map(|change| change.facet_value.into()),
+        )?;
+
+        self.inner.add_or_delete_level(wtxn)
+    }
+}
+
+impl FacetsUpdateIncrementalInner {
+    /// WARNING: `touched_children` must be sorted in **reverse** lexicographic order.
+    fn find_touched_parents(
+        &self,
+        wtxn: &mut RwTxn,
+        child_level: u8,
+        mut touched_children: impl Iterator<Item = Vec<u8>>,
+    ) -> Result<()> {
+        let mut touched_parents = vec![];
+        let Some(parent_level) = child_level.checked_add(1) else { return Ok(()) };
+        let parent_level_left_bound: FacetGroupKey<&[u8]> =
+            FacetGroupKey { field_id: self.field_id, level: parent_level, left_bound: &[] };
+
+        let mut last_parent: Option<Vec<u8>> = None;
+
+        for child in &mut touched_children {
+            if !valid_facet_value(&child) {
+                continue;
+            }
+
+            if let Some(last_parent) = &last_parent {
+                if child.as_slice() >= last_parent.as_slice() {
+                    self.compute_parent_group(wtxn, child_level, child)?;
```

Comment:
I am not sure why we need to call `compute_parent_group`, is it in the case the current child is a totally new one and should be registered in the parent?

Reply:
`compute_parent_group` considers its argument (here, the `child`) as a group, and recomputes it from its own children.

When the level argument to `compute_parent_group` is `0`, then the call is a no-op.

Otherwise, we need to compute the group before discarding the `child`, otherwise some groups will not be refreshed.

I hope this clarifies this part of the algorithm, available for a call if more information is needed